ouster_ros/src/motion_correction.cpp

```diff
@@ -0,0 +1,160 @@
+  }
+}
+
+void MotionCorrection::addPointCloudToQueue(
+    const std::shared_ptr<pcl::PointCloud<PointOS1>>& pointcloud_ptr) {
+  if (!pointcloud_list_.empty() &&
+      pointcloud_ptr->header.stamp < pointcloud_list_.back()->header.stamp) {
+    ROS_ERROR_STREAM(
+        "Detected jump backwards in time, resetting pointcloud list");
+    pointcloud_list_.clear();
+  }
+
+  if (pointcloud_list_.size() > max_pointcloud_queue_length_) {
+    ROS_WARN_STREAM("There are now over "
+                    << max_pointcloud_queue_length_
+                    << " pointclouds waiting to be transformed, "
+                       "removing oldest.");
+    pointcloud_list_.pop_front();
+  }
+
+  pointcloud_list_.push_back(pointcloud_ptr);
+}
+
+bool MotionCorrection::processNextQueuedPointcloud(
+    pcl::PointCloud<pcl::PointXYZI>* pointcloud) {
+  while (!pointcloud_list_.empty()) {
+    const InterpolationStatus status =
+        transformPoints(*pointcloud_list_.front(), pointcloud);
+
+    if (status == InterpolationStatus::MATCHED) {
+      pointcloud_list_.pop_front();
+      return true;
+    }
+    if ((status == InterpolationStatus::AFTER_LAST) ||
+        (status == InterpolationStatus::NOT_ENOUGH_POINTS)) {
+      return false;
+    }
+    if (status == InterpolationStatus::BEFORE_FIRST) {
+      pointcloud_list_.pop_front();
+    }
+  }
+  return false;
+}
+
+MotionCorrection::InterpolationStatus MotionCorrection::transformPoints(
+    const pcl::PointCloud<PointOS1>& pointcloud_in,
+    pcl::PointCloud<pcl::PointXYZI>* pointcloud_out) {
+  pointcloud_out->header = pointcloud_in.header;
+  pointcloud_out->reserve(pointcloud_in.size());
+
+  ros::Time input_time;
+  const uint64_t first_point_stamp =
+      pointcloud_in.header.stamp + pointcloud_in.front().time_offset_us;
+  const uint64_t last_point_stamp =
+      pointcloud_in.header.stamp + pointcloud_in.back().time_offset_us;
+  pointcloud_out->header.stamp =
+      pointcloud_in.header.stamp + (pointcloud_in.front().time_offset_us +
+                                    pointcloud_in.back().time_offset_us) /
+                                       2;
+
+  if (transformation_list_.size() < 2) {
+    return InterpolationStatus::NOT_ENOUGH_POINTS;
+  }
+  if (transformation_list_.front().stamp > first_point_stamp) {
+    return InterpolationStatus::BEFORE_FIRST;
+  }
+  if (transformation_list_.back().stamp < last_point_stamp) {
+    return InterpolationStatus::AFTER_LAST;
+  }
+
+  while (cloud_it_->stamp < pointcloud_out->header.stamp) {
+    ++cloud_it_;
+  }
+
+  Interpolator interpolator(*std::prev(cloud_it_), *cloud_it_);
+  const Transformation transformation_pointcloud_inv =
+      interpolator(pointcloud_out->header.stamp).inverse();
+
+  AlignedList<TransformationStamped>::const_iterator previous_point_it =
+      cloud_it_;
+  uint64_t previous_point_stamp = 0ul;
+  Eigen::Affine3f pcl_transform;
+
+  for (const PointOS1& point : pointcloud_in) {
```

Comment:
Isn't taking blocks of 64 points only valid if you assume that every point gives a valid return? Though I agree it could be more efficient to do in blocks.

Reply:
In some implementations, it may be better to keep all points and just assign 0 range for points that have no valid range. Points without valid range still have valid intensity information. This lets us keep a fixed 64x2048 structured representation, which is useful for things like transforming blocks. It is also useful for things like associating pixels in the intensity image with 3D points; masking out parts of the field of view that are on the robot; using a lookup table to convert between raw range readings and xyz cartesian coordinates; and so on.

For example, `PointOS1Cloud` in [viz.cpp#L265](https://github.com/ouster-lidar/ouster_example/blob/master/ouster_viz/src/viz.cpp#L265) is a struct which is always fixed size, allowing ranges to be converted to xyz cartesian coordinates using a fixed size lookup table, which is very efficient.